### PR TITLE
Improved parsing infrastructure

### PR DESCRIPTION
**Pull request by toinehartman**
Parsing happens on a separate `Thread`. Also, parsing code is moved from
several `programminglife.model` classes to a dedicated
`programminglife.parser.GraphParser` class.

### CODE REVIEW

File: src/main/java/programminglife/parser/GraphParser.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+            if (n != null && n.getParents().isEmpty()) {
+                this.getGraph().getRootNodes().add(n);
+            }
+        }
+    }
+
+    public Graph getGraph() {
+        return graph;
+    }
+}
```
Comment (YDutchy) [inline src/main/java/programminglife/parser/GraphParser.java]: Good to see all the parsing being moved to its own class.

File: src/test/java/programminglife/gui/GuiControllerTest.java [hunk 2]
```diff
@@ -35,7 +35,6 @@
         stage.show();
         stage.toFront();
         File f = new File(getClass().getResource("/test.gfa").getPath());
-        Graph g = Graph.parse(f);
```
Comment (YDutchy) [inline src/test/java/programminglife/gui/GuiControllerTest.java]: Redundant for now, as it is handled in the pull request from the GuiConrollerTest

File: src/test/java/programminglife/parser/GraphParserTest.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        graphParser.parseSegment(nodeLine);
+
+        Node node = graphParser.getGraph().getNode(6);
+
+        assertEquals(6, node.getIdentifier());
+        assertEquals("C", node.getSequence());
+        assertEquals(0, node.getParents().size());
+        assertEquals(0, node.getChildren().size());
+    }
+}
```
Comment (YDutchy) [inline src/test/java/programminglife/parser/GraphParserTest.java]: Good addition of immediately having tests for the newly added class.

File: src/main/java/programminglife/gui/controller/GuiController.java [hunk 4]
```diff
@@ -81,17 +84,29 @@ private void initialize() {
...
+            Graph graph = (Graph) arg;
             this.graphController.setGraph(graph);
-            consoleView.appendText(String.format("The graph has %d nodes\n", graph.size()));
 
-            disableGraphUIElements(false);
+            if (graph != null) {
+                disableGraphUIElements(false);
+            } else {
+                disableGraphUIElements(true);
+            }
```
Comment (MeAmAnUsername) [inline src/main/java/programminglife/gui/controller/GuiController.java]: This if is not necessary, just use `disableGraphUIElements(graph == null)`

File: src/main/java/programminglife/parser/GraphParser.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        try {
+            System.out.printf("%s Parsing Graph on separate Thread", Thread.currentThread());
+            parse(this.verbose);
+            this.setChanged();
+            this.notifyObservers(this.graph);
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+        } catch (UnknownTypeException e) {
+            e.printStackTrace();
+        }
```
Comment (MeAmAnUsername) [inline src/main/java/programminglife/parser/GraphParser.java]: Please actually do something when an exception is thrown (Notify observers that parsing failed?)

File: src/main/java/programminglife/gui/controller/GuiController.java [hunk 6]
```diff
@@ -88,28 +88,30 @@ public void openFile(File file) throws FileNotFoundException, UnknownTypeExcepti
...
+        if (o instanceof GraphParser) {
+            if (arg instanceof Graph) {
+                Graph graph = (Graph) arg;
+                this.graphController.setGraph(graph);
+
+                if (graph != null) {
+                    disableGraphUIElements(false);
+                } else {
+                    disableGraphUIElements(true);
+                }
```
Comment (MeAmAnUsername) [inline src/main/java/programminglife/gui/controller/GuiController.java]: You could still just replace this by `disableGraphUIElements(graph == null)`

File: src/main/java/programminglife/parser/GraphParser.java (hunk 5 above)
Comment (MeAmAnUsername) [inline src/main/java/programminglife/parser/GraphParser.java]: Fixed, looks better.

File: src/test/java/programminglife/parser/GraphParserTest.java [hunk 7]
```diff
@@ -70,4 +73,35 @@ public void parseSegmentTest() {
...
+    public void runTestSuccess() {
+        graphParser.addObserver(this);
+        graphParser.run();
+    }
+
+    @Test
+    public void runTestFailure() {
+        try {
+            faultyGraphParser.addObserver(this);
+            faultyGraphParser.run();
```
Comment (MeAmAnUsername) [inline src/test/java/programminglife/parser/GraphParserTest.java]: Also put `fail();` after this to make sure it won't pass when it doesn't throw an exception.

**Comment by toinehartman** [comment]
@MeAmAnUsername everything should be fixed now.

File: src/test/java/programminglife/parser/GraphParserTest.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        } catch (RuntimeException re) {
+            assertEquals(UnknownTypeException.class, re.getCause().getClass());
```
Comment (MeAmAnUsername) [inline src/test/java/programminglife/parser/GraphParserTest.java]: You could also use `throw re.getCause()` and then expect it in the `@Test` above, but that is more of a design choice.
I am in favor of that as that makes it clear that we expect an Exception (as opposed to having a kind of obscure assertEquals(...) on the class of the Exception)

Comment (MeAmAnUsername) [inline src/test/java/programminglife/parser/GraphParserTest.java]: Test...

**Comment by MeAmAnUsername** [comment]
Note: github hates me, but the last two reviews are *not* outdated (yet)

File: src/test/java/programminglife/parser/GraphParserTest.java (hunk 8 above)
Comment (toinehartman) [inline src/test/java/programminglife/parser/GraphParserTest.java]: Unfortunately, it seems we cannot.
screenshot 2017-05-16 21 45 03

Comment (MeAmAnUsername) [inline src/test/java/programminglife/parser/GraphParserTest.java]: Have you tried adding `throws FooException` to the method signature?

**Pull request merged**